### ISSUE TITLE
[Bug]: Accessibility. Elements in virtual machines list are not announced when navigating with keyboard and screenreader enabled

### ISSUE THREAD

**Issue by sevapopov2** [VirtualBox/virtualbox#39]
### Version

main

### Host OS Type

Windows

### Host Architecture

X64

### Guest OS Type

Not applicable.

### Guest Architecture

Not applicable

### Component

Accessibility

### What happened?

Screenreader doesn't announce the elements in focus in virtual machines list when navigating with up and down arrow keys.

## Preconditions
Screenreader enabled.
There are 1 or more virtual machines present in the list.

## Steps to reproduce

1. Navigate to virtual machines list.
2. Press up and down arrow keys to move between the list elements.

## Actual behavior
Screenreader doesn't announce the element in focus when navigating to it with arrow keys in virtual machines list.

## Expected behavior
Screenreader announces elements in focus.

## Reproduced on
NVDA 2024.2.
Windows 11 24H2 (26100.4061).
VirtualBox 7.1.10-169112.

## Logs
[Logs.zip](https://github.com/user-attachments/files/20698067/Logs.zip)

### Did you upload *all* of your necessary log files, screenshots, etc.?

- [x] Yes, I've uploaded all pertinent files to this issue.

**Comment by Dsen-orcl** [comment]
Hello and thanks for that one report!
That took a bit of time but finally I found the reason of NVDA unable to work with our machine tree. The issue will be fixed in subsequent 7.2.x. If that will be possible I'll backport these changes to 7.1 as well, we will see.

**Comment by sevapopov2** [comment]
Hello,
@Dsen-orcl Thank you a lot for paying attention to the issue and fixing it!
Will wait until the next version when fix is released and report other accessibility issues!
Is it needed to send links to new accessibility issues in VirtualBox users mailing list or simply reporting it in Github is enough for it to be seen?
Thank you!

**Comment by Dsen-orcl** [comment]
> Hello, [@Dsen-orcl](https://github.com/Dsen-orcl) Thank you a lot for paying attention to the issue and fixing it! Will wait until the next version when fix is released and report other accessibility issues!

BTW, I can provide you with Windows test-build as well if you'd like to test fix before the release.

> Is it needed to send links to new accessibility issues in VirtualBox users mailing list or simply reporting it in Github is enough for it to be seen? Thank you!

My preference would certainly be a Github record*, i.e. separate Github issue per each problem you've met.

**Comment by sevapopov2** [comment]
> BTW, I can provide you with Windows test-build as well if you'd like to test fix before the release.

Yes, please do! I will be glad to test the fix and report if something is wrong.

**Comment by Dsen-orcl** [comment]
> Yes, please do! I will be glad to test the fix and report if something is wrong.

Do you mind testing development snapshot for now (which means VBox of 7.2 version)? I know original report is for the 7.1 version, I'll prepare 7.1 testbuild later as soon as all the required changes backported.

Here is the Testbuilds link: https://www.virtualbox.org/wiki/Testbuilds <= please look for development revision 170703.

**Comment by Dsen-orcl** [comment]
For now I'm working on keyboard navigation and keyboard focus frames for various places around VirtualBox Manager window and Global Preferences / VM Settings dialogs. These accessibility aspects are known to be broken and some reporters were telling me about that already.

Though I know that NVDA screen-reader isn't working well with our GUI in many places and that will be the next thing I'll be working on (so far only the place mentioned in this Github defect should be fixed by the testbuild). Previously macOS VoiceOver screen-reader was our main priority, now it's time to extend NVDA coverage as well.

**Comment by sevapopov2** [comment]
> For now I'm working on keyboard navigation and keyboard focus frames for various places around VirtualBox Manager window and Global Preferences / VM Settings dialogs. These accessibility aspects are known to be broken and some reporters were telling me about that already.

Yes, that is true.
For example there are issues currently that:

1. For some reason even if you choose option category in VM settings with up and down arrow keys you still get all settings when tabbing separated with tabs. It is really confusing. I think it became not that long time ago.
2. Currently there is an issue that you cannot tab around VM settings with tab key, there is a keyboard trap that only moves between search field, OK, Cancel and Apply buttons.
3. Network settings also have issues like when adding a port forwarding rule for SSH it is tricky.

I can report all these things in Github if needed, I think it is better even if you know about that so they will be documented.

> Though I know that NVDA screen-reader isn't working well with our GUI in many places and that will be the next thing I'll be working on (so far only the place mentioned in this Github defect should be fixed by the testbuild). Previously macOS VoiceOver screen-reader was our main priority, now it's time to extend NVDA coverage as well.

Great to see that VoiceOver for MacOS also got atention!
When I tried to use VirtualBox for MacOS several years ago it's accessibility was really good in what I used.
However, Windows is the most accessible and popular operating system for blind users and it is awesome that now you have systematic aproach to accessibility for it!
Currently, using VirtualBox Manager with NVDA is really tricky and requires time to find workarounds and ways to access different things and elements.
Since I am an accessibility tester, I am an experienced screenreader user and I know all that and found ways and workarounds to perform things I need in virtual machine settings, but I will really look forward to more improvements!
Do you need screen recordings with defects demonstrations or text descriptions are enough?
By the way, it is great that VirtualBox now has it's Github repository, it was needed to register in it's own bug tracker before and as far as I could tell (maybe I am wrong but I experienced that) accessibility issues were not really noticed and only got atention when I subscribed to users mailing list and reported about them.
Thank you for your work!

**Comment by Dsen-orcl** [comment]
> For example there are issues currently that:
> 
> 1. For some reason even if you choose option category in VM settings with up and down arrow keys you still get all settings when tabbing separated with tabs. It is really confusing. I think it became not that long time ago.

I don't think I can do much regarding 1st issue, because when you choose option category - it just scrolls the view to corresponding section, but TABbing through the view still uses overall tab-order (including widgets from other sections) as all widgets are placed into one view now. I'd proposed to create a separate defect for this one and describe how would you see the expected behavior there.

> 2. Currently there is an issue that you cannot tab around VM settings with tab key, there is a keyboard trap that only moves between search field, OK, Cancel and Apply buttons.

2nd issue should be fixed in the testbuild already.

> 3. Network settings also have issues like when adding a port forwarding rule for SSH it is tricky.

Please create separate defect regarding 3rd one issue, this place can be improved I think.

**Comment by Dsen-orcl** [comment]
> Great to see that VoiceOver for MacOS also got atention!
> When I tried to use VirtualBox for MacOS several years ago it's accessibility was really good in what I used.

Right, macOS VoiceOver was our target accessibility platform initially, as years ago (when we've started to work on accessibility) VoiceOver was much more stable and easy to sustain comparing to NVDA. I should note that implementing accessibility interfaces for VoiceOver and NVDA is not always cross-platform effort, they often require different aspects to be implemented for each of them. Currently we have intention to iteratively improve NVDA support with your help.

**Comment by Dsen-orcl** [comment]
> Do you need screen recordings with defects demonstrations or text descriptions are enough?

Initially (if issue is reproducible) text description should be enough.

> By the way, it is great that VirtualBox now has it's Github repository, it was needed to register in it's own bug tracker before and as far as I could tell (maybe I am wrong but I experienced that) accessibility issues were not really noticed and only got atention when I subscribed to users mailing list and reported about them.

That's mostly because we were trying to improve accessibility support for macOS first of all. Because initially NVDA support was not possible at all as Windows hardening was preventing it from work with our product, so we've decided to postpone it until it's possible. The problem was solved few years ago but higher priorities for other tasks delayed improvements in the area for a moment.

**Comment by Dsen-orcl** [comment]
Hello @sevapopov2 ,

I'm going to mark this defect (about announcing VM tree items via NVDA) for closing as it's fixed for me, can you confirm that as well?

Please also mention here every other defect you've created on the basis of our discussion above. I'll assign them to myself.

BTW, I have implemented NVDA support and keyboard navigation for VM tree, Global and Machine toolbars and section selector list of Global preferences / VM settings. Items should be navigateable now via keyboard and announced by NVDA properly.

Currently I'm working on NVDA accessibility support for all kinds of list/tree/table widgets we have. Some of fixes will be a part of nearest minor release, some will come later.

**Comment by sevapopov2** [comment]
> I'm going to mark this defect (about announcing VM tree items via NVDA) for closing as it's fixed for me, can you confirm that as well?

What version should I confirm it on?
Main 7.1?
Also, should I report other defects on main stable version or to install the latest test builds and report on them?

**Comment by Dsen-orcl** [comment]
> What version should I confirm it on?

Ah, right, no 7.1 backport so far. As we're discussing whether it worth backporting the fix to 7.1 as the changes are large enough. So it's for 7.2 testbuild for now, the one I've uploaded two weeks ago. And it will be a part of nearest 7.2 minor release, but still unclear about 7.1.

**Comment by Dsen-orcl** [comment]
> Also, should I report other defects on main stable version or to install the latest test builds and report on them?

7.2 preferably

**Comment by Dsen-orcl** [comment]
> Currently I'm working on NVDA accessibility support for all kinds of list/tree/table widgets we have. Some of fixes will be a part of nearest minor release, some will come later.

Right, forgot to mention that all other fixes goes only to 7.2. I mean there is still question about backporting fix for this only defect to 7.1, but rest of fixes will go to 7.2 only. Too many changes to backport to relatively stable 7.1.

**Comment by sevapopov2** [comment]
> Right, forgot to mention that all other fixes goes only to 7.2. I mean there is still question about backporting fix for this only defect to 7.1, but rest of fixes will go to 7.2 only. Too many changes to backport to relatively stable 7.1.

Thank you for pointing that. Yes, you can close this issue with machines announcement in the list, it is no longer reproducible.
I checked on version 7.2.3-170703.
Here are my 2 questions:

1. Is it normal that virtual machines created in VirtualBox 7.1 don't work in 7.2? I never really tested it that much so not sure if it is normal or not. My virtual machines created on 7.1 don't launch.
2. When I created [last issue](https://github.com/virtualbox/virtualbox/issues/266) at this moment of writing, I chose main/trunk version in versions list, however I reproduced it on build 7.2.3-170703. The reason why I chose main/trunk because the latest version present there that time was 7.2.2. What should I do in this situation, what version to choose in the list? Latest available or to create an issue and insert the version by editing it?

Thank you!

**Comment by Dsen-orcl** [comment]
> Yes, you can close this issue with machines announcement in the list, it is no longer reproducible.
> I checked on version 7.2.3-170703.

Thanks for the confirmation!

> 1. Is it normal that virtual machines created in VirtualBox 7.1 don't work in 7.2? I never really tested it that much so not sure if it is normal or not. My virtual machines created on 7.1 don't launch.

That can be the case but every occurance have to be analyzed separately, so that probably deserves a separate defect with the description and VBox.log file attached. Our guys usually able to pinpoint and explain such (version mismatch) issues fast enough. Also, having this as a separate defect useful for a case when somebody else meet the same issue in future.

> 2. When I created last issue at this moment of writing, I chose main/trunk version in versions list, however I reproduced it on build 7.2.3-170703. The reason why I chose main/trunk because the latest version present there that time was 7.2.2. What should I do in this situation, what version to choose in the list? Latest available or to create an issue and insert the version by editing it?

I would choose main/trunk in such case and would separately mentioned (in the defect description) that it is still reproducible with the recent testbuild of the specified version/revision.

**Comment by sevapopov2** [comment]
@Dsen-orcl I created 2 new issues regarding network port forwarding settings #289 and #290. Thank you!

**Comment by sevapopov2** [comment]
> > For example there are issues currently that:
> > 
> > 1. For some reason even if you choose option category in VM settings with up and down arrow keys you still get all settings when tabbing separated with tabs. It is really confusing. I think it became not that long time ago.
> 
> I don't think I can do much regarding 1st issue, because when you choose option category - it just scrolls the view to corresponding section, but TABbing through the view still uses overall tab-order (including widgets from other sections) as all widgets are placed into one view now. I'd proposed to create a separate defect for this one and describe how would you see the expected behavior there.

Created issue #295 describing how navigation in settings should work with screenreader in my opinion.
Also created issue #293.
